File: moderate/sum4.lua
function numzero(s, c, z)
  if c == 0 then
    return (z == 0) and 1 or 0
  elseif #s < c or z + c * s[1] > 0 or z + c * s[#s] < 0 then
    return 0
  end
  local t = {}
  for i = 1, #s-1 do t[#t+1] = s[i+1] end
  return numzero(t, c-1, z + s[1]) + numzero(t, c, z)
end

for line in io.lines(arg[1]) do
  local s = {}
  for i in line:gmatch("[-]?%d+") do
    s[#s+1] = tonumber(i)
  end
  table.sort(s)
  print(numzero(s, 4, 0))
end

File: moderate/sum4.hs
import System.Environment (getArgs)
import Data.List.Split (splitOn)
import Data.List (sort)

numzero      :: Int -> Int -> [Int] -> Int
numzero 0 0 _ = 1
numzero 0 _ _ = 0
numzero c z s | length s < c       = 0
              | z + c * head s > 0 = 0
              | z + c * last s < 0 = 0
              | otherwise          = numzero (c - 1) (z + head s) (tail s) + numzero c z (tail s)

main :: IO ()
main = do
    [inpFile] <- getArgs
    input <- readFile inpFile
    putStr . unlines . map (show . numzero 4 0 . sort . map read . splitOn ",") $ lines input
